Extend expr_initializer to support byte-wise initialization

peterschrammel opened the pull request:
Each byte of the expression is initialized to the given initialization expression, if possible.

This building block will be required for the shadow memory implementation.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Refactor expr_initializer to take initialization expression

We can now pass in 0 and nondet, and can then extend it to
take more general initialization expressions.

peterschrammel committed:
Extend expr_initializer to support byte-wise initialization

Each byte of the expression is initialized to the given
initialization expression.

This building block will be required for the shadow
memory implementation.

tautschnig commented:
Given I don't have enough context just yet the following may be a stupid idea, apologies for that: couldn't you instead use `byte_extract(array_of<char[sizeof(target-type)]>(init_value), 0, target-type)`?

peterschrammel commented:
> Given I don't have enough context just yet the following may be a stupid idea, apologies for that: couldn't you instead use byte_extract(array_of<char[sizeof(target-type)]>(init_value), 0, target-type)?

That may not be as stupid as it sounds. Let me try. If it works then the entire `expr_initializer` class can be replaced by this single statement.

kroening commented:
> > Given I don't have enough context just yet the following may be a stupid idea, apologies for that: couldn't you instead use byte_extract(array_of<char[sizeof(target-type)]>(init_value), 0, target-type)?
> 
> That may not be as stupid as it sounds. Let me try. If it works then the entire `expr_initializer` class can be replaced by this single statement.

Please keep in mind that Goto programs are meant to be consumed by a broad variety of analysis methods. Virtually all but symex will choke on the above.

esteffin commented on src/util/expr_initializer.cpp:
```diff
@@ -307,5 +323,69 @@ optionalt<exprt> nondet_initializer(
+{
+  if(output_type.id() == ID_unsignedbv || output_type.id() == ID_signedbv)
+  {
+    const size_t size =
+      to_bitvector_type(output_type).get_width() / config.ansi_c.char_width;
```
This should check that `to_bitvector_type(output_type).get_width()` is divisible by `config.ansi_c.char_width` otherwise it may be smaller!!!

peterschrammel replied:
We should probably fall back to a cast if it is smaller (and this behaviour needs to be documented).

esteffin replied:
Changed

esteffin commented on src/util/expr_initializer.cpp:
```diff
@@ -307,5 +323,69 @@ optionalt<exprt> nondet_initializer(
+/// Builds an expression of the given output type with each of its bytes
+/// initialized to the given initialization expression.
+/// Integer bitvector types are currently supported.
+/// For unsupported types the initialization expression is casted to the
+/// output type.
```
Should this be an invariant instead?

peterschrammel replied:
No, we need to deal with all output types (casting is the fallback behaviour).

esteffin replied:
Left as it was

esteffin commented on src/util/expr_initializer.cpp:
```diff
@@ -307,5 +323,69 @@ optionalt<exprt> nondet_initializer(
+/// \return The built expression
+exprt duplicate_per_byte(const exprt &init_byte_expr, const typet &output_type)
+{
+  if(output_type.id() == ID_unsignedbv || output_type.id() == ID_signedbv)
+  {
+    const size_t size =
+      to_bitvector_type(output_type).get_width() / config.ansi_c.char_width;
```
Why not a more modern?

```suggestion
  if(
    const auto bv_type =
      type_try_dynamic_cast<integer_bitvector_typet>(output_type))
  {
    const size_t size = bv_type->get_width() / config.ansi_c.char_width;
```

esteffin replied:
Changed

esteffin commented:
> > > Given I don't have enough context just yet the following may be a stupid idea, apologies for that: couldn't you instead use byte_extract(array_of<char[sizeof(target-type)]>(init_value), 0, target-type)?
> > 
> > 
> > That may not be as stupid as it sounds. Let me try. If it works then the entire `expr_initializer` class can be replaced by this single statement.
> 
> Please keep in mind that Goto programs are meant to be consumed by a broad variety of analysis methods. Virtually all but symex will choke on the above.

I agree. We should create a goto that can be consumed by analysis methods other than symex.
An option can be to run `lower_byte_operators` to the output of the `byte_extract(array_of<char[sizeof(target-type)]>(init_value), 0, target-type)` call, but it seems not so clean compared to what is in this PR (just an extension of the already existing `expr_initializert`).

esteffin reviewed (changes requested):
Looks good although it needs unit tests + a few minor comments

esteffin commented:
Rebased to the latest `origin/develop`.

Commit:
Improved checks and fixed duplicate_per_byte

Function `duplicate_per_byte` now has a PRECONDITION checking that the
`init_byte_expr` is either a boolean or a bitvector of maximum size 8.

Also the computation of the number of duplication correctly accounts the
case when the destination is not divisible by 8. In this case the input
is duplicated one extra (including the sign bit if any) time and then
truncated to the output size.

Commit:
Added tests for duplicate_per_byte in the BV case

Commit:
Test for zero_initializer and expr_initializer

esteffin commented:
PR rebased, added tests and ready to review.

peterschrammel reviewed (commented):
LGTM
